File: contrib/PC/psptools/psptool/pbp.py
import struct

PBP_HEADER_MAGIC = 0x50425000


def is_pbp(data):
    return struct.unpack('<I', data[:0x4])[0] == PBP_HEADER_MAGIC


class PBP:
    def __init__(self, data):
        (self.magic, self.version, sfo_offset, icon0_offset,
         icon1_offset, pic0_offset, pic1_offset,
         snd0_offset, prx_offset, psar_offset) = struct.unpack('<IIIIIIIIII', data[:0x28])

        self.sfo = data[sfo_offset:icon0_offset]
        self.icon0 = data[icon0_offset:icon1_offset]
        self.icon1 = data[icon1_offset:pic0_offset]
        self.pic0 = data[pic0_offset:pic1_offset]
        self.pic1 = data[pic1_offset:snd0_offset]
        self.snd0 = data[snd0_offset:prx_offset]
        self.prx = data[prx_offset:psar_offset]
        self.psar = data[psar_offset:]

    def pack(self):
        return struct.pack('<IIIIIIIIII', self.magic, self.version,
                           0x28,
                           0x28+len(self.sfo),
                           0x28+len(self.sfo)+len(self.icon0),
                           0x28+len(self.sfo)+len(self.icon0)+len(self.icon1),
                           0x28+len(self.sfo)+len(self.icon0) +
                           len(self.icon1)+len(self.pic0),
                           0x28+len(self.sfo)+len(self.icon0)+len(self.icon1) +
                           len(self.pic0)+len(self.pic1),
                           0x28+len(self.sfo)+len(self.icon0)+len(self.icon1) +
                           len(self.pic0)+len(self.pic1)+len(self.snd0),
                           0x28+len(self.sfo)+len(self.icon0)+len(self.icon1) +
                           len(self.pic0)+len(self.pic1) +
                           len(self.snd0)+len(self.prx)
                           ) + self.sfo + self.icon0 + self.icon1 + self.pic0 + self.pic1 + self.snd0 + self.prx + self.psar

File: contrib/PC/pbp/pbp.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

typedef uint32_t u32;

typedef struct
{
    u32 magic;
    u32 version;
    u32 param_offset;
    u32 icon0_offset;
    u32 icon1_offset;
    u32 pic0_offset;
    u32 pic1_offset;
    u32 snd0_offset;
    u32 elf_offset;
    u32 psar_offset;
} PBPHeader;

int main(){

    FILE* fp = fopen("EBOOT.PBP", "rb");
    
    
    PBPHeader header;
    fread(&header, 1, sizeof(header), fp);
    
    u32 size;
    
    size = header.psar_offset - header.elf_offset;
    if (size){
        FILE* out = fopen("simple.prx", "wb");
        void* data = malloc(size);
        fseek(fp, header.elf_offset, SEEK_SET);
        fread(data, 1, size, fp);
        fwrite(data, 1, size, out);
        free(data);
        fclose(out);
    }
    
    size = header.icon0_offset - header.param_offset;
    if (size){
        FILE* out = fopen("param.sfo", "wb");
        void* data = malloc(size);
        fseek(fp, header.elf_offset, SEEK_SET);
        fread(data, 1, size, fp);
        fwrite(data, 1, size, out);
        free(data);
        fclose(out);
    }
    
    fseek(fp, 0, SEEK_END);
    size = ftell(fp) - header.psar_offset;
    if (size){
        FILE* out = fopen("data.psar", "wb");
        void* data = malloc(size);
        fseek(fp, header.elf_offset, SEEK_SET);
        fread(data, 1, size, fp);
        fwrite(data, 1, size, out);
        free(data);
        fclose(out);
    }

    fclose(fp);

    return 0;

}
